### PR TITLE
Fix wrap asis output

### LINKED ISSUE #5506: Quarto fails when `x` in `knitr::asis_output(x)` has `length(x) != 1`
### Bug description

Quarto fails when `x` in `knitr::asis_output(x)` has `length(x) != 1`.

See PR #5505 for a solution.

#### Minimal reproducible example

Example qmd:

````md
---
title: Fix asis_output
format: html
engine: knitr
---

`asis_output` with one string:

```{r}
knitr::asis_output("* one\n* two")
```

`asis_output` with two strings:

```{r}
knitr::asis_output(c("* one\n", "* two\n"))
```

`asis_output` with zero strings:
```{r}
knitr::asis_output(character(0))
```
````

The second codeblock generates a warning:

    In grepl("^<\\w+[ >]", x) && grepl("<\\/\\w+>\\s*$", x) :
      'length(x) = 2 > 1' in coercion to 'logical(1)'

While the third codeblock produces an error:

    Error in if (grepl("^<\\w+[ >]", x) && grepl("<\\/\\w+>\\s*$", x) && !grepl("^<div class=\"kable-table\">",  : 
      missing value where TRUE/FALSE needed
    Calls: .main ... sew -> <Anonymous> -> lapply -> FUN -> wrap_asis_output

<details><summary>Command output</summary>

    $ quarto render demo_qmd.qmd
    
    
    processing file: demo_qmd.qmd
      |....................................................| 100% (unnamed-chunk-3)Quitting from lines 21-22 (demo_qmd.qmd) 
                                                                                                                
    Error in if (grepl("^<\\w+[ >]", x) && grepl("<\\/\\w+>\\s*$", x) && !grepl("^<div class=\"kable-table\">",  : 
      missing value where TRUE/FALSE needed
    Calls: .main ... sew -> <Anonymous> -> lapply -> FUN -> wrap_asis_output
    In addition: Warning message:
    In grepl("^<\\w+[ >]", x) && grepl("<\\/\\w+>\\s*$", x) :
      'length(x) = 2 > 1' in coercion to 'logical(1)'
    Execution halted

</details>

#### Expected output

I expect the intermediary markdown to look like this (but then in html and not in markdown):

<details><summary>demo_qmd.md</summary>

````md
# Fix asis_output

`asis_output` with one string:

```{r}
knitr::asis_output("* one\n* two")
```

* one
* two

`asis_output` with two strings:

```{r}
knitr::asis_output(c("* one\n", "* two\n"))
```

* one
* two

`asis_output` with zero strings:
```{r}
knitr::asis_output(character(0))
```
````

</details>

#### Comparison to Rmarkdown

With RMarkdown, this code works just fine:

<details><summary>demo_rmd.Rmd</summary>

````md
---
title: Fix asis_output
output:
  html_document
---

`asis_output` with one string:

```{r}
knitr::asis_output("* one\n* two")
```

`asis_output` with two strings:

```{r}
knitr::asis_output(c("* one\n", "* two\n"))
```

`asis_output` with zero strings:
```{r}
knitr::asis_output(character(0))
```
````
</details>

<details><summary>Command output</summary>

    $ Rscript -e 'rmarkdown::render("demo_rmd.Rmd")'
    
    
    processing file: demo_rmd.Rmd
                                                                                                                
    output file: demo_rmd.knit.md
    
    /usr/bin/pandoc +RTS -K512m -RTS demo_rmd.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output demo_rmd.html --lua-filter /home/rcannood/R/x86_64-redhat-linux-gnu-library/4.2/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /home/rcannood/R/x86_64-redhat-linux-gnu-library/4.2/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --section-divs --template /home/rcannood/R/x86_64-redhat-linux-gnu-library/4.2/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmpxfSAAT/rmarkdown-str9883135bff75b.html 
    
    Output created: demo_rmd.html

</details>

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [ ] Please document the operating system you're running. If on Linux, please provide the specific distribution.
- [ ] Please provide the output of `quarto check` so we know which version of quarto and its dependencies you're running.

### PR DESCRIPTION
**Pull request by rcannood**
## Description

Closes #5506

## Checklist

I have (if applicable):

- [x] filed a [contributor agreement](https://github.com/quarto-dev/quarto-cli/blob/main/CONTRIBUTING.md).
- [x] referenced the GitHub issue this PR closes
- [x] updated the appropriate changelog

### CODE REVIEW

**Comment by rcannood** [comment]
What's going on in this PR? :P

If it helps, I can explain the major changes proposed by this PR.

The `wrap_asis_output` function has two arguments, `options` and `x`. Up until now, it was assumed that the `x` (which is the value when a user uses `knitr::asis_output(x)` has exactly length 1, whereas x can actually be a character vector of length n. This PR changes the `wrap_asis_output` implementation to account for this behaviour.

**Comment by cscheid** [comment]
> What's going on with reviewing this PR? :P

Everyone has a full plate! We appreciate the patience.

**Comment by rcannood** [comment]
Hah. I don't mind waiting, but it's just funny to see the suggested reviewers change every day or so.

**Comment by cscheid** [comment]
> the suggested reviewers change every day or so.

This is a github annoyance. Every click you make on the UI, even by accident, causes a change that is interpreted as an action that is logged on the issue... I apologize if you get notified by this!

**Review by cderv — CHANGES_REQUESTED** [review]
Sorry for the delay @rcannood. I was onto other topics, and wanted to make some time to test this correctly. I was under the impression it was not as easy as it looked like. 

It seems to me merging this PR would break such example 

````markdown
---
title: Fix asis_output
format: html
keep-md: true
---

```{r}
library(kableExtra)
dt <- mtcars[1:5, 1:6]
dt %>%
  kbl() %>%
  kable_styling()
```
````

image

There are too many `cell-output-display` wrapping, maybe because replacing `paste0()` by `c()` is not exactly the same. 

So making the change to fix #5506 requires a bit more thinking, or at least more testing to get the correct fix.

**Comment by rcannood** [comment]
Hey @cderv!

Thanks for looking into this.

I hadn't spotted other places where I should have changed `paste0()` into `c()` to get this to work. However, in this PR I now just added a check that collapses the string together if it's length isn't 1.

What do you think would be a better solution? Refactoring `wrap_asis_output()` such that `length(x) != 1` is allowed, or simply collapsing the string when that condition occurs?

I'm also working on adding a smoke test (rcannood/quarto-cli#1), though I don't know the syntax, so I don't know whether I'm doing it right ;)

Robrecht

File: tests/docs/smoke-all/2023/05/31/issue-5506.qmd
Comment (cderv) [inline tests/docs/smoke-all/2023/05/31/issue-5506.qmd:10]: The syntax is 
````yaml
ensureFileRegexMatches:    
    - ["regex_to_match", "another_regex to match"]
    - ["regex_to_not_match", "another regex_to_not_match"]
````

So `ensureFileRegexMatches` takes 
* First a list of regex that should be match - if they are not match, then the test will error
* Second a list of regex that should not be match - if they are the test will error

It is indeed not documented so you couldn't know without reading the `.ts` file. 

Some tips for future PR: Using `html` as a format allows you to use `ensureHtmlElements` which works similarly by using CSS selectors - this can sometime be easier than regex.

**Review by cderv — CHANGES_REQUESTED** [review]
I just left a comment to help you fix the test. As you are adding test, (thank you!) you could also add one with the **kable** example I shared above.

File: tests/docs/smoke-all/2023/05/31/issue-5506.qmd
Comment (rcannood) [reply tests/docs/smoke-all/2023/05/31/issue-5506.qmd]: Thanks for the information!

I tried to have a go at this but am not at all familiar with the unit testing notation. When I render the test qmd manually it looks good.

Would you be willing to take a look at what I'm doing wrong with the unit test?

Comment (cderv) [reply tests/docs/smoke-all/2023/05/31/issue-5506.qmd]: Sure. I'll make time later to look at it. I am at a conference this week, so a bit slower than usual, but I'll have a look.  

I think you just need to escape properly the regex and / or use some CSS selector instead of regex.

Comment (rcannood) [reply tests/docs/smoke-all/2023/05/31/issue-5506.qmd]: Thanks for the tip; I managed to solve the issue. Since I managed to solve fix the unit test, I'll consider this issue resolved and will request a re-review :)

Enjoy your conference!

**Comment by cderv** [comment]
Yes I'll get to it - it is on my list. I am assigned to it, with review requested, so be sure I'll get to it at some point. I am just onto another thing right now. 

Also, something can seem simple but it can have side effects that are less simple. This often depends on where in the code base it is. And this is at the root of the R processing. So there could be some side effect to some content and we want to absolutely prevent that - I want to verify that, and it is less quick that a simple very scoped fix. 

Are you blocked in some way by #5505 ? is this very high priority for you ?

**Comment by rcannood** [comment]
I'm no longer blocked by this as I can paste my character vector before passing it to `knitr::asis_output()`.

However, the error messages and warning messages mentioned in #5506 are very unclear and it took me some time to figure out what the issue was and how to fix it. With this PR I hope to ensure that nobody else encounters these cryptic errors or warnings.

:bow: :)

**Comment by rcannood** [comment]
Kind reminder for this PR, as some of my quarto projects are complaining about this bug.

I think the proposed patch is relatively simple in that the current code was written such that it only expects `x` to be of length 1. My proposed solution simply concatenates `x` together if `length(x) != 1`.

**Comment by cderv** [comment]
Thanks for the reminder ! I'll have a new look at this 

> as some of my quarto projects are complaining about this bug.

Can you share with a real life example of this `knitr::asis_output()` usage you have ? 
I would like to check something related to previous behavior encountered (e.g [difference with `#| output: asis` and `knitr::asis_output()` usage](https://github.com/quarto-dev/quarto-cli/discussions/6874#discussioncomment-7070893))

**Comment by rcannood** [comment]
We're using it here:

https://github.com/openproblems-bio/website/blob/981de26295e3e6d6c55add2ca9e86b347283eefe/results/_include/_method_descriptions.qmd#L25

Be warned that this is quite a large repository with a lot of R & Python code. I could make a smaller version of this repo by removing the `/documentation` directory.

**Comment by rcannood** [comment]
A small update: the code that was previously working now doesn't work anymore in R 4.3 since it now returns an error if the condition has length > 1:

```r
x <- c(TRUE, FALSE)
if (x) { cat("foo\n") }
```

    Error in if (x) { : the condition has length > 1

**Comment by cderv** [comment]
Oh ok, thank you. i'll add a test for this then;

**Comment by cderv** [comment]
Ok it fails indeed. This is mixed up with other failing test we have on main right now;

**Comment by cderv** [comment]
Let's always collapse. I believe it is clearer than just testing for length one. Thanks

**Review by cderv — APPROVED** [review]
I'll merge once main is fixed

**Comment by rcannood** [comment]
Alright, awesome! Thanks a lot :)

**Comment by rcannood** [comment]
:partying_face:

**Comment by cderv** [comment]
Thanks @rcannood and sorry for the delay on this. Thank you for your patience !

**Pull request merged**
